jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/ResponseListeners.java

```diff
@@ -452,43 +454,35 @@ private void onDemandCallback()
 
         private void registerFailure(Throwable failure)
         {
-            while (true)
+            try (AutoLock ignored = lock.lock())
             {
-                long encoded = counters.get();
-                int failures = AtomicBiInteger.getHi(encoded) + 1;
-                int demands = AtomicBiInteger.getLo(encoded);
+                failures++;
+                if (failures == listeners.size())
+                    originalContentSource.fail(failure);
+
                 if (demands == listeners.size() - failures)
```

Comment:
I'd prefer `if (demands + failures == listeners.size())`.

Also, must guarantee that we do not count 1 demand and 1 failure for the same ContentSource instance.